codes/c/chapter_array_and_linkedlist/list.c

```diff
@@ -0,0 +1,187 @@
+/**
+ * File: list.c
+ * Update Time: 2023-02-12
+ * Author: Guanngxu (446678850@qq.com)
+ */
+
+#include "../include/include.h"
+
+// C 语言中没有 list，用数组实现 list
+struct List {
+    int *items;       // 数组（存储列表元素）
+    int capacity;    // 列表容量
+    int size;        // 列表大小
+    int extendRatio; // 列表每次扩容的倍数
+};
+
+typedef struct List List;
+
+/* 前置声明 */
+void extendCapacity(List *list);
+
+/* 构造方法 */
+List *newList() {
+    List *list = malloc(sizeof(List));
+    list->capacity = 10;
+    list->items = malloc(sizeof(int) * list->capacity);
+    list->size = 0;
+    list->extendRatio = 2;
+    return list;
+}
+
+/* 析构方法 */
+void delList(List *list) {
+    free(list->items);
+    free(list);
+}
+
+/* 访问元素 */
+int get(List *list, int index) {
+    assert(index >= 0 && index < list->size);
+    return list->items[index];
+}
+
+/* 更新元素 */
+void set(List *list, int index, int num) {
+    assert(index >= 0 && index < list->size);
+    list->items[index] = num;
+}
+
+/* 尾部添加元素 */
+void add(List *list, int num) {
+    if (list->size == list->capacity) {
+        extendCapacity(list); // 扩容
+    }
+    list->items[list->size] = num;
+    list->size++;
+}
+
+/* 中间插入元素 */
+void insert(List *list, int index, int num) {
+    assert(index >= 0 && index < list->size);
+    for (int i = list->size; i > index; --i) {
+        list->items[i] = list->items[i - 1];
+    }
+    list->items[index] = num;
+    list->size++;
+}
+
+/* 删除元素 */
+// 由于引入了 stdio.h ，此处无法使用 remove 关键词
+// 详见 https://github.com/krahets/hello-algo/pull/244#discussion_r1067863888
+int removeItem(List *list, int index) {
+    assert(index >= 0 && index < list->size);
+    int num = list->items[index];
+    for (int i = index; i < list->size - 1; i++) {
+        list->items[i] = list->items[i + 1];
+    }
+    list->size--;
+    return num;
+}
+
+/* 列表扩容 */
+void extendCapacity(List *list) {
+    // 先分配空间
+    int newCapacity = list->capacity * list->extendRatio;
+    int *extend = (int *) malloc(sizeof(int) * newCapacity);
+    int *temp = list->items;
+
+    // 拷贝旧数据到新数据
+    for (int i = 0; i < list->size; i++)
+        extend[i] = list->items[i];
+
+    // 释放旧数据
+    free(temp);
+
+    // 更新新数据
+    list->items = extend;
+    list->capacity = newCapacity;
+}
+
+/* 拼接列表 */
+List *append(List *listA, List *listB) {
+    List *ret = newList();
+    for (int i = 0; i < listA->size; i++)
+    {
+        add(ret, listA->items[i]);
+    }
+    for (int i = 0; i < listB->size; i++)
+    {
+        add(ret, listB->items[i]);
+    }
+    return ret;
+}
+
+/* 排序比较函数 */
+int cmp(const void *a,const void *b) {
```

Comment:
Less is more. Keeping consistent with other language implementations.

Reply:
`cmp` is required in `qsort`.